src/specBuilder/marks/markUtils.ts

```diff
@@ -273,3 +300,127 @@ export const getVoronoiPath = (children: MarkChildElement[], dataSource: string,
 		},
 	],
 });
+/**
+ * the signal that triggers the opacity ease in and out
+ * @param opacityValue
+ * @returns { signal: string}
+ */
+//TODO: Add tests
+export const getHighlightOpacityAnimationValue = (opacityValue: { signal: string } | { value: number }): { signal: string }  => {
+	if ('signal' in opacityValue) {
+		return { signal: `max(1-rscColorAnimation, ${opacityValue.signal} / ${HIGHLIGHT_CONTRAST_RATIO})` }
+	}
+	return { signal: `max(1-rscColorAnimation, ${opacityValue.value} / ${HIGHLIGHT_CONTRAST_RATIO})`}
+};
+/**
+ * animation opacity rules for charts that highlight from series ID
+ * @param opacityValue
+ * @returns ProductionRule<NumericValueRef>
+ */
+//TODO: Add tests
+export const getSeriesAnimationOpacityRules = (
+	opacityValue?: { signal: string } | { value: number },
+): ProductionRule<NumericValueRef> => {
+	if (!opacityValue) {
+		opacityValue = DEFAULT_OPACITY_RULE;
+	}
```

Comment:
A neat way to do this is with a nullish coalescing operator (??). It will check the first condition, and if it resolves to true, that condition will become the value. Otherwise, the second condition will become the value. 

```suggestion
	opacityValue = opacityValue ?? DEFAULT_OPACITY_RULE;
```

Reply:
https://javascript.info/nullish-coalescing-operator